### ISSUE TITLE
Stopping continous_recognition_async after mic is stopped for certain time

### ISSUE THREAD

**Issue by sids07** [Azure-Samples/cognitive-services-speech-sdk#1947]
Hello, i was trying recognize_once_async method and it worked perfectly. But main issue was it only listens to audio for max 15 sec as mentioned on docs. So, i was trying  start_continuous_recognition_async so that it can work for longer audio. And came to know that method stop_continuous_recognition_async can be used for stop listening. And i was curious if somehow we can write logic which hear silence for certain time and stop continous recognition.

Thanks.

**Comment by glharper** [comment]
@sids07 Thanks for using the speech SDK and writing this issue up. For continuous recognition you can set the end silence timeout property in milliseconds on the speech config object. In python, this looks like:
```
    # set end silence timeout to 3s
    speech_config.set_property(speechsdk.PropertyId.SpeechServiceConnection_EndSilenceTimeoutMs, "3000")
```

You should received a canceled event after 3 seconds of silence, and can stop recognition then.

**Comment by sids07** [comment]
@glharper thanks for the response but i have already tried setting different properties such as EndSilenceTimeoutMs, InitialSilenceTimeoutMs, SegmentationSilenceTimeoutms. But neither of them are throwing canceled event. Below is the code i am trying to run.
`
def speech_recognize_continuous_async_from_microphone():
    """performs continuous speech recognition asynchronously with input from microphone"""
    speech_config = speechsdk.SpeechConfig(subscription=os.environ.get("SPEECH_KEY"), region=os.environ.get("SPEECH_REGION"))
    speech_config.set_property(speechsdk.PropertyId.SpeechServiceConnection_EndSilenceTimeoutMs, "500")
    speech_config.set_property(speechsdk.PropertyId.SpeechServiceConnection_InitialSilenceTimeoutMs, "500")
    speech_config.set_property(speechsdk.PropertyId.Speech_SegmentationSilenceTimeoutMs, "500")

    # The default language is "en-us".
    speech_recognizer = speechsdk.SpeechRecognizer(speech_config=speech_config)

    done = False

    def recognizing_cb(evt: speechsdk.SpeechRecognitionEventArgs):
        print('RECOGNIZING: {}'.format(evt))

    def recognized_cb(evt: speechsdk.SpeechRecognitionEventArgs):
        print('RECOGNIZED: {}'.format(evt))

    def stop_cb(evt: speechsdk.SessionEventArgs):
        """callback that signals to stop continuous recognition"""
        print('CLOSING on {}'.format(evt))
        nonlocal done
        done = True

    # Connect callbacks to the events fired by the speech recognizer
    speech_recognizer.recognizing.connect(recognizing_cb)
    speech_recognizer.recognized.connect(recognized_cb)
    speech_recognizer.session_stopped.connect(stop_cb)
    speech_recognizer.canceled.connect(stop_cb)

    # Perform recognition. `start_continuous_recognition_async asynchronously initiates continuous recognition operation,
    # Other tasks can be performed on this thread while recognition starts...
    # wait on result_future.get() to know when initialization is done.
    # Call stop_continuous_recognition_async() to stop recognition.
    result_future = speech_recognizer.start_continuous_recognition_async()

    result_future.get()  # wait for voidfuture, so we know engine initialization is done.
    print('Continuous Recognition is now running, say something.')

    while not done:
        # No real sample parallel work to do on this thread, so just wait for user to type stop.
        # Can't exit function or speech_recognizer will go out of scope and be destroyed while running.
        print('type "stop" then enter when done')
        stop = input()
        if (stop.lower() == "stop"):
            print('Stopping async recognition.')
            speech_recognizer.stop_continuous_recognition_async()
            break
    
    speech_recognizer.stop_continuous_recognition_async()
`

**Comment by glharper** [comment]
@pankopon Could you take a look at this?

**Comment by pankopon** [comment]
@sids07 Please see attached an example. Timeouts are reported using `ResultReason.NoMatch` (not `ResultReason.Canceled`) with `NoMatchReason.InitialSilenceTimeout` or `NoMatchReason.EndSilenceTimeout`.
[timeout.zip](https://github.com/Azure-Samples/cognitive-services-speech-sdk/files/11521167/timeout.zip)

Unfortunately there has been some service issues regarding silence timeouts recently, and depending on your region the timeouts may or may not be reported properly - for this reason, the example code contains also a "workaround" to check for empty recognition results.

Also, you should not try using very small values, they may be ignored by the service or not work as intended. Try the values used in the example and only increase them if necessary. (`EndSilenceTimeoutMs` in particular is better left as is.) If you intend to stop recognition based on silence i.e. no more speech expected, you probably should wait for several seconds anyway.

The silence timeouts in continuous recognition (should) work so that in the very beginning before any speech is recognized, there can be multiple `InitialSilenceTimeoutMs` periods of silence, then each recognition result can be followed by `EndSilenceTimeoutMs` plus multiple `InitialSilenceTimeoutMs` periods of silence. If the service does not report `EndSilenceTimeoutMs` separately then the first silence timeout period immediately after a recognition result will be `EndSilenceTimeoutMs` + `InitialSilenceTimeoutMs`.

An alternative to these silence timeout settings would be to use a push stream for Speech SDK input and read audio data from the source (microphone) and write to the push stream while measuring the length of silence in the application, but that would obviously mean more work for the application to do.

**Comment by sids07** [comment]
@pankopon thanks for the help. I have tried the example code you provided, everythings seems to be logical and should have worked. But as you mentioned there are issues regarding silence timeouts, and i think i am facing the same issue. Silence Timeouts are not throwing any results, so the program waits for recognizing empty text (i.e. around 30 seconds which is way too much) then only the program ends listening.

Will also look into push stream for speech sdk. But i was curious if there is any workarounds such that we can make silence timout issue work.

**Comment by pankopon** [comment]
@sids07 Did you run the code as is with the included example settings for initial and end silence timeouts (4500 + 1500 ms)? Testing with `westcentralus` region I can get results like the following where the timeout after the last spoken input (the time diff between `[10:40:56.964] RECOGNIZED: This is the last speech.` and `[10:41:03.191] RECOGNIZED:`) corresponds to this setting. What is the region you are using?
```
[10:40:49.162] ENTER recognize_speech_from_microphone
[10:40:49.245] Session started: SessionEventArgs(session_id=2f87b9f0d7d64b929267803440a7be50)
[10:40:52.467] Recognizing: testing testing
[10:40:53.102] RECOGNIZED: Testing. Testing.
[10:40:55.351] Recognizing: do you still
[10:40:55.952] Recognizing: this is the last
[10:40:56.451] Recognizing: this is the last speech
[10:40:56.964] RECOGNIZED: This is the last speech.
[10:41:03.191] RECOGNIZED:
[10:41:03.191] Closing because of presumed silence timeout
[10:41:03.199] Session stopped: SessionEventArgs(session_id=2f87b9f0d7d64b929267803440a7be50)
[10:41:03.201] EXIT recognize_speech_from_microphone
14.040 s
```
@mahilleb-msft @fmegen FYI, issues with silence timeouts in continuous recognition.

**Comment by sids07** [comment]
@pankopon  Yes i did run the code as it was on included examples for initial and end silence timeouts (4500 + 1500 ms). My region is `centralindia` . Below is the log i received.

` 
[09:40:11.067] ENTER recognize_speech_from_microphone

[09:40:11.105] Session started: SessionEventArgs(session_id=c5dbf18d033242079e9ead115107c165)
[09:40:12.791] Recognizing: i know
[09:40:13.295] Recognizing: i know i know
[09:40:15.323] RECOGNIZED: I know, I know.
[09:41:23.142] RECOGNIZED: 
[09:41:23.142] Closing because of presumed silence timeout
[09:41:23.148] Session stopped: SessionEventArgs(session_id=c5dbf18d033242079e9ead115107c165)
[09:41:23.148] EXIT recognize_speech_from_microphone `

**Comment by pankopon** [comment]
@sids07 Update: I've found out that continuous recognition now seems to require `OutputFormat.Detailed` in order to make silence timeouts to be reported properly (I'm checking if this is really intentional on the service side). In other words, add the following setting
```
speech_config.output_format = speechsdk.OutputFormat.Detailed
```
With this I can get results like, a silence timeout when there is no speech
```
[17:00:48.385] ENTER recognize_speech_from_microphone
[17:00:48.442] Session started: SessionEventArgs(session_id=efef6896322d4d4dad0f5751d0f94e2a)
[17:00:53.812] NOMATCH: NoMatchReason.InitialSilenceTimeout
[17:00:53.813] Closing because of InitialSilenceTimeout
[17:00:53.820] Session stopped: SessionEventArgs(session_id=efef6896322d4d4dad0f5751d0f94e2a)
[17:00:53.821] EXIT recognize_speech_from_microphone
5.437 s
```
and after some speech
```
[17:00:56.703] ENTER recognize_speech_from_microphone
[17:00:56.748] Session started: SessionEventArgs(session_id=13aaa12009e6406a82464b19fda0de40)
[17:00:59.374] Recognizing: testing
[17:00:59.870] RECOGNIZED: Testing.
[17:01:06.117] NOMATCH: NoMatchReason.InitialSilenceTimeout
[17:01:06.117] Closing because of InitialSilenceTimeout
[17:01:06.123] Session stopped: SessionEventArgs(session_id=13aaa12009e6406a82464b19fda0de40)
[17:01:06.124] EXIT recognize_speech_from_microphone
9.421 s
```
(the timeout after recognized speech is really EndSilenceTimeoutMs + InitialSilenceTimeoutMs)

Please try the same (and note that if you are using a microphone, any additional sounds may affect when the timeout actually occurs).

**Comment by pankopon** [comment]
Internal work item ref. 5273024 for tracking the issue of missing silence timeouts in continuous recognition. Until resolved, an additional setting is required to enable timeouts as described in the previous comment.

**Comment by pankopon** [comment]
The service issue with missing silence timeouts in continuous recognition has been fixed. I verified that the timeouts are now reported as expected by default and it is not necessary to set `OutputFormat.Detailed`.

**Comment by pankopon** [comment]
Closing the issue as resolved, please open a new item if more support is needed.
